File: app/ts/common/availability.ts
import debugModule from 'debug';
import { getDate } from './conversions';
import { toJSON } from './parser';
import { settings as appSettings, Settings } from './settings';
import { checkPatterns } from './whoiswrapper/patterns';
import { predict as aiPredict } from '../ai/availabilityModel';

const debug = debugModule('common.whoisWrapper');
let settings: Settings = appSettings;

export interface WhoisResult {
  domain?: string;
  status?: string;
  registrar?: string;
  company?: string;
  creationDate?: string | undefined;
  updateDate?: string | undefined;
  expiryDate?: string | undefined;
  whoisreply?: string;
  whoisJson?: Record<string, unknown>;
}

interface CheckContext {
  resultsText: string;
  resultsJSON: Record<string, unknown>;
  domainParams: WhoisResult;
  controlDate: string | undefined;
}

type CheckFn = (ctx: CheckContext) => boolean;

const AVAILABLE_PATTERNS: string[] = [
  'No match for domain',
  '- No Match',
  'NO MATCH:',
  'No match for',
  'No match',
  'No matching record.',
  'Nincs talalat',
  'Status: AVAILABLE',
  'Status:             AVAILABLE',
  'Status:         available',
  'Status: free',
  'Status: Not Registered',
  'query_status: 220 Available',
  'This domain name has not been registered',
  'The domain has not been registered',
  'This query returned 0 objects',
  'domain name not known in',
  'registration status: available',
  'Object does not exist',
  'The queried object does not exist',
  'Not Registered -',
  'is available for registration',
  'is available for purchase',
  'DOMAIN IS NOT A REGISTERD',
  'No such domain',
  'No_Se_Encontro_El_Objeto',
  'Domain unknown',
  'No information available about domain name',
  'is not valid!'
];

const AVAILABLE_FUNCTIONS: CheckFn[] = [
  (ctx) =>
    ctx.domainParams.expiryDate !== undefined &&
    ctx.controlDate !== undefined &&
    Date.parse(ctx.domainParams.expiryDate) - Date.parse(ctx.controlDate) < 0,
  (ctx) =>
    ctx.resultsText.includes(' is free') &&
    ctx.domainParams.whoisreply !== undefined &&
    ctx.domainParams.whoisreply.length < 50,
  (ctx) =>
    ctx.resultsText.includes('whois.nic.bo') &&
    ctx.domainParams.whoisreply !== undefined &&
    ctx.domainParams.whoisreply.length < 55,
  (ctx) => ctx.resultsText.includes('Error.') && ctx.resultsText.includes('SaudiNIC')
];

const UNAVAILABLE_PATTERNS: string[] = [
  'Domain Status:ok',
  'Expiration Date:',
  'Expiry Date:',
  'Status: connect',
  'Changed:',
  'organisation: Internet Assigned Numbers Authority'
];

const UNAVAILABLE_FUNCTIONS: CheckFn[] = [
  (ctx) => Object.prototype.hasOwnProperty.call(ctx.resultsJSON, 'domainName'),
  (ctx) => Object.keys(ctx.resultsJSON).length > 5
];

const ERROR_CHECKS: { result: string; strings?: string[]; fn?: CheckFn }[] = [
  { result: 'error:nocontent', fn: (ctx) => ctx.resultsText === null || ctx.resultsText === '' },
  {
    result: 'error:unauthorized',
    strings: ['You  are  not  authorized  to  access or query our Whois']
  },
  {
    result: 'error:ratelimiting',
    strings: [
      'IP Address Has Reached Rate Limit',
      'Too many connection attempts',
      'Your request is being rate limited',
      'Your query is too often.',
      'Your connection limit exceeded.'
    ]
  },
  { result: 'error:unretrivable', strings: ['Could not retrieve Whois data'] },
  {
    result: 'error:forbidden',
    strings: ['si is forbidden', 'Requests of this client are not permitted']
  },
  { result: 'error:reservedbyregulator', strings: ['reserved by aeDA Regulator'] },
  { result: 'error:unregistrable', strings: ['third-level domains may not start with'] },
  {
    result: 'error:replyerror',
    strings: [
      'error ',
      'error',
      'Error',
      'ERROR:101:',
      'Whois lookup error',
      'can temporarily not be answered',
      'Invalid input'
    ],
    fn: (ctx) =>
      Object.prototype.hasOwnProperty.call(ctx.resultsJSON, 'error') ||
      Object.prototype.hasOwnProperty.call(ctx.resultsJSON, 'errno')
  }
];

export function isDomainAvailable(
  resultsText: string,
  resultsJSON?: Record<string, unknown>
): string {
  const { lookupAssumptions: assumptions } = settings;

  if (settings.ai.enabled) {
    try {
      const aiRes = aiPredict(resultsText);
      if (aiRes === 'available' || aiRes === 'unavailable') return aiRes;
    } catch (e) {
      debug(`AI prediction failed: ${e}`);
    }
  }

  const patternResult = checkPatterns(resultsText, resultsJSON);
  const defaultResult = assumptions.unparsable ? 'available' : 'error:unparsable';
  if (patternResult !== defaultResult) return patternResult;

  if (!resultsJSON) resultsJSON = toJSON(resultsText) as Record<string, unknown>;

  const domainParams = getDomainParameters(null, null, null, resultsJSON, true);
  const controlDate = getDate(new Date());

  if (resultsText.includes('Uniregistry') && resultsText.includes('Query limit exceeded')) {
    return assumptions.uniregistry ? 'unavailable' : 'error:ratelimiting';
  }

  const ctx: CheckContext = {
    resultsText,
    resultsJSON,
    domainParams,
    controlDate
  };

  for (const p of AVAILABLE_PATTERNS) {
    if (resultsText.includes(p)) return 'available';
  }
  for (const fn of AVAILABLE_FUNCTIONS) {
    if (fn(ctx)) return 'available';
  }

  for (const p of UNAVAILABLE_PATTERNS) {
    if (resultsText.includes(p)) return 'unavailable';
  }
  for (const fn of UNAVAILABLE_FUNCTIONS) {
    if (fn(ctx)) return 'unavailable';
  }

  for (const check of ERROR_CHECKS) {
    if (check.fn && check.fn(ctx)) return check.result;
    if (check.strings) {
      for (const s of check.strings) if (resultsText.includes(s)) return check.result;
    }
  }

  return assumptions.unparsable ? 'available' : 'error:unparsable';
}

export function getDomainParameters(
  domain: string | null,
  status: string | null,
  resultsText: string | null,
  resultsJSON: Record<string, unknown>,
  isAuxiliary = false
): WhoisResult {
  const results: WhoisResult = {};

  results.domain = domain ?? undefined;
  results.status = status ?? undefined;
  results.registrar = resultsJSON.registrar as string | undefined;
  results.company =
    (resultsJSON.registrantOrganization as string | undefined) ||
    (resultsJSON.registrant as string | undefined) ||
    (resultsJSON.adminName as string | undefined) ||
    (resultsJSON.ownerName as string | undefined) ||
    (resultsJSON.contact as string | undefined) ||
    (resultsJSON.name as string | undefined);
  results.creationDate = getDate(
    (resultsJSON.creationDate as string | boolean | Date | undefined) ||
      (resultsJSON.createdDate as string | boolean | Date | undefined) ||
      (resultsJSON.created as string | boolean | Date | undefined) ||
      (resultsJSON.registered as string | boolean | Date | undefined) ||
      (resultsJSON.registeredOn as string | boolean | Date | undefined)
  );
  results.updateDate = getDate(
    (resultsJSON.updatedDate as string | boolean | Date | undefined) ||
      (resultsJSON.lastUpdated as string | boolean | Date | undefined) ||
      (resultsJSON.UpdatedDate as string | boolean | Date | undefined) ||
      (resultsJSON.changed as string | boolean | Date | undefined) ||
      (resultsJSON.lastModified as string | boolean | Date | undefined) ||
      (resultsJSON.lastUpdate as string | boolean | Date | undefined)
  );
  results.expiryDate = getDate(
    (resultsJSON.expires as string | boolean | Date | undefined) ||
      (resultsJSON.registryExpiryDate as string | boolean | Date | undefined) ||
      (resultsJSON.expiryDate as string | boolean | Date | undefined) ||
      (resultsJSON.registrarRegistrationExpirationDate as string | boolean | Date | undefined) ||
      (resultsJSON.expire as string | boolean | Date | undefined) ||
      (resultsJSON.expirationDate as string | boolean | Date | undefined) ||
      (resultsJSON.expiresOn as string | boolean | Date | undefined) ||
      (resultsJSON.paidTill as string | boolean | Date | undefined)
  );
  results.whoisreply = resultsText ?? undefined;
  results.whoisJson = resultsJSON;

  return results;
}

export default {
  isDomainAvailable,
  getDomainParameters
};

File: src-tauri/src/availability.rs
use serde::{Serialize, Deserialize};
use crate::parser::parse_raw_data;

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
pub enum DomainStatus {
    #[serde(rename = "available")]
    Available,
    #[serde(rename = "unavailable")]
    Unavailable,
    #[serde(rename = "error")]
    Error,
    #[serde(rename = "error:unparsable")]
    ErrorUnparsable,
    #[serde(rename = "error:ratelimiting")]
    ErrorRateLimiting,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct WhoisParams {
    pub domain: Option<String>,
    pub status: Option<DomainStatus>,
    pub registrar: Option<String>,
    pub company: Option<String>,
    #[serde(rename = "creationDate")]
    pub creation_date: Option<String>,
    #[serde(rename = "updateDate")]
    pub update_date: Option<String>,
    #[serde(rename = "expiryDate")]
    pub expiry_date: Option<String>,
    pub whoisreply: Option<String>,
}

const AVAILABLE_PATTERNS: &[&str] = &[
    "No match for domain",
    "- No Match",
    "NO MATCH:",
    "No match for",
    "No match",
    "No matching record.",
    "Nincs talalat",
    "Status: AVAILABLE",
    "Status: free",
    "Status: Not Registered",
    "query_status: 220 Available",
    "This domain name has not been registered",
    "The domain has not been registered",
    "This query returned 0 objects",
    "domain name not known in",
    "registration status: available",
    "Object does not exist",
    "The queried object does not exist",
    "Not Registered -",
    "is available for registration",
    "is available for purchase",
    "DOMAIN IS NOT A REGISTERD",
    "No such domain",
    "No_Se_Encontro_El_Objeto",
    "Domain unknown",
    "No information available about domain name",
    "is not valid!"
];

const UNAVAILABLE_PATTERNS: &[&str] = &[
    "Domain Status:ok",
    "Expiration Date:",
    "Expiry Date:",
    "Status: connect",
    "Changed:",
    "organisation: Internet Assigned Numbers Authority"
];

pub fn is_domain_available(results_text: &str) -> DomainStatus {
    for p in AVAILABLE_PATTERNS {
        if results_text.contains(p) {
            return DomainStatus::Available;
        }
    }

    for p in UNAVAILABLE_PATTERNS {
        if results_text.contains(p) {
            return DomainStatus::Unavailable;
        }
    }

    if results_text.contains("Uniregistry") && results_text.contains("Query limit exceeded") {
        return DomainStatus::ErrorRateLimiting;
    }

    DomainStatus::Unavailable // Default to unavailable if no clear available pattern matches
}

pub fn get_domain_parameters(
    domain: Option<String>,
    status: Option<DomainStatus>,
    results_text: String,
) -> WhoisParams {
    let results_json = parse_raw_data(&results_text);
    
    let registrar = results_json.get("registrar").cloned();
    let company = results_json.get("registrantOrganization")
        .or_else(|| results_json.get("registrant"))
        .or_else(|| results_json.get("adminName"))
        .or_else(|| results_json.get("ownerName"))
        .or_else(|| results_json.get("contact"))
        .or_else(|| results_json.get("name"))
        .cloned();
        
    let creation_date = results_json.get("creationDate")
        .or_else(|| results_json.get("createdDate"))
        .or_else(|| results_json.get("created"))
        .or_else(|| results_json.get("registered"))
        .or_else(|| results_json.get("registeredOn"))
        .cloned();

    let update_date = results_json.get("updatedDate")
        .or_else(|| results_json.get("lastUpdated"))
        .or_else(|| results_json.get("UpdatedDate"))
        .or_else(|| results_json.get("changed"))
        .or_else(|| results_json.get("lastModified"))
        .or_else(|| results_json.get("lastUpdate"))
        .cloned();

    let expiry_date = results_json.get("expires")
        .or_else(|| results_json.get("registryExpiryDate"))
        .or_else(|| results_json.get("expiryDate"))
        .or_else(|| results_json.get("registrarRegistrationExpirationDate"))
        .or_else(|| results_json.get("expire"))
        .or_else(|| results_json.get("expirationDate"))
        .or_else(|| results_json.get("expiresOn"))
        .or_else(|| results_json.get("paidTill"))
        .cloned();

    WhoisParams {
        domain,
        status,
        registrar,
        company,
        creation_date,
        update_date,
        expiry_date,
        whoisreply: Some(results_text),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    // ── is_domain_available ──────────────────────────────────────────────

    #[test]
    fn test_is_domain_available() {
        assert_eq!(is_domain_available("No match for domain example.com"), DomainStatus::Available);
        assert_eq!(is_domain_available("Status: free"), DomainStatus::Available);
        assert_eq!(is_domain_available("Domain Status:ok"), DomainStatus::Unavailable);
        assert_eq!(is_domain_available("Expiration Date: 2025-01-01"), DomainStatus::Unavailable);
        assert_eq!(is_domain_available("Uniregistry Query limit exceeded"), DomainStatus::ErrorRateLimiting);
    }

    #[test]
    fn test_availability_edge_cases() {
        // Empty text
        assert_eq!(is_domain_available(""), DomainStatus::Unavailable);

        // Conflicting patterns (Available pattern usually takes precedence in simple matching)
        assert_eq!(is_domain_available("No match for domain but Expiration Date: 2020"), DomainStatus::Available);

        // No patterns matching
        assert_eq!(is_domain_available("some random text with no keywords"), DomainStatus::Unavailable);

        // Case sensitivity (our patterns are currently exact, we should check if they should be case-insensitive)
        // The original JS used text.includes(p), which is case-sensitive.
        assert_eq!(is_domain_available("STATUS: AVAILABLE"), DomainStatus::Unavailable); // Case mismatch
    }

    #[test]
    fn test_all_available_patterns() {
        let test_cases = vec![
            "No match for domain foobar.com",
            "- No Match",
            "NO MATCH:",
            "No match for something",
            "No match",
            "No matching record.",
            "Nincs talalat",
            "Status: AVAILABLE",
            "Status: free",
            "Status: Not Registered",
            "query_status: 220 Available",
            "This domain name has not been registered",
            "The domain has not been registered",
            "This query returned 0 objects",
            "domain name not known in xyz",
            "registration status: available",
            "Object does not exist",
            "The queried object does not exist",
            "Not Registered - example.com",
            "example.com is available for registration",
            "is available for purchase",
            "DOMAIN IS NOT A REGISTERD",
            "No such domain",
            "No_Se_Encontro_El_Objeto",
            "Domain unknown",
            "No information available about domain name",
            "is not valid!",
        ];
        for text in test_cases {
            assert_eq!(
                is_domain_available(text),
                DomainStatus::Available,
                "Expected Available for: {}",
                text
            );
        }
    }

    #[test]
    fn test_all_unavailable_patterns() {
        let test_cases = vec![
            "Domain Status:ok\nMore data",
            "Expiration Date: 2030-01-01",
            "Expiry Date: 2030-01-01",
            "Status: connect",
            "Changed: 2024-01-01",
            "organisation: Internet Assigned Numbers Authority",
        ];
        for text in test_cases {
            assert_eq!(
                is_domain_available(text),
                DomainStatus::Unavailable,
                "Expected Unavailable for: {}",
                text
            );
        }
    }

    #[test]
    fn test_uniregistry_rate_limit() {
        assert_eq!(
            is_domain_available("Uniregistry  blah blah Query limit exceeded  blah"),
            DomainStatus::ErrorRateLimiting
        );
    }

    #[test]
    fn test_uniregistry_without_limit() {
        // Has Uniregistry but not rate limit text
        assert_eq!(
            is_domain_available("Uniregistry Registrar"),
            DomainStatus::Unavailable
        );
    }

    // ── DomainStatus serde ───────────────────────────────────────────────

    #[test]
    fn test_domain_status_serialization() {
        assert_eq!(serde_json::to_string(&DomainStatus::Available).unwrap(), "\"available\"");
        assert_eq!(serde_json::to_string(&DomainStatus::Unavailable).unwrap(), "\"unavailable\"");
        assert_eq!(serde_json::to_string(&DomainStatus::Error).unwrap(), "\"error\"");
        assert_eq!(serde_json::to_string(&DomainStatus::ErrorUnparsable).unwrap(), "\"error:unparsable\"");
        assert_eq!(serde_json::to_string(&DomainStatus::ErrorRateLimiting).unwrap(), "\"error:ratelimiting\"");
    }

    #[test]
    fn test_domain_status_deserialization() {
        let avail: DomainStatus = serde_json::from_str("\"available\"").unwrap();
        assert_eq!(avail, DomainStatus::Available);
        let err: DomainStatus = serde_json::from_str("\"error:ratelimiting\"").unwrap();
        assert_eq!(err, DomainStatus::ErrorRateLimiting);
    }

    // ── get_domain_parameters ────────────────────────────────────────────

    #[test]
    fn test_get_domain_parameters_basic() {
        let text = "Registrar: GoDaddy LLC\nCreation Date: 2020-01-01\nExpiry Date: 2030-01-01";
        let params = get_domain_parameters(
            Some("example.com".into()),
            Some(DomainStatus::Unavailable),
            text.into(),
        );
        assert_eq!(params.domain, Some("example.com".into()));
        assert_eq!(params.registrar, Some("GoDaddy LLC".into()));
        assert_eq!(params.creation_date, Some("2020-01-01".into()));
        assert_eq!(params.expiry_date, Some("2030-01-01".into()));
    }

    #[test]
    fn test_get_domain_parameters_company_chain() {
        // Should try registrantOrganization first, then registrant, etc.
        let text = "Registrant Organization: ACME Corp";
        let params = get_domain_parameters(None, None, text.into());
        assert_eq!(params.company, Some("ACME Corp".into()));
    }

    #[test]
    fn test_get_domain_parameters_company_fallback_admin() {
        let text = "Admin Name: John Doe";
        let params = get_domain_parameters(None, None, text.into());
        assert_eq!(params.company, Some("John Doe".into()));
    }

    #[test]
    fn test_get_domain_parameters_update_date_variants() {
        let text = "Updated Date: 2024-06-15";
        let params = get_domain_parameters(None, None, text.into());
        assert_eq!(params.update_date, Some("2024-06-15".into()));

        let text2 = "Last Updated: 2024-01-01";
        let params2 = get_domain_parameters(None, None, text2.into());
        assert_eq!(params2.update_date, Some("2024-01-01".into()));
    }

    #[test]
    fn test_get_domain_parameters_expiry_variants() {
        let text1 = "Expires: 2030-12-31";
        let params1 = get_domain_parameters(None, None, text1.into());
        assert_eq!(params1.expiry_date, Some("2030-12-31".into()));

        let text2 = "Registry Expiry Date: 2028-06-01";
        let params2 = get_domain_parameters(None, None, text2.into());
        assert_eq!(params2.expiry_date, Some("2028-06-01".into()));

        let text3 = "paid-till: 2027-03-15";
        let params3 = get_domain_parameters(None, None, text3.into());
        assert_eq!(params3.expiry_date, Some("2027-03-15".into()));
    }

    #[test]
    fn test_get_domain_parameters_empty_text() {
        let params = get_domain_parameters(None, None, "".into());
        assert!(params.registrar.is_none());
        assert!(params.company.is_none());
        assert!(params.creation_date.is_none());
        assert!(params.update_date.is_none());
        assert!(params.expiry_date.is_none());
    }

    #[test]
    fn test_get_domain_parameters_whoisreply_included() {
        let text = "Some raw WHOIS data";
        let params = get_domain_parameters(None, None, text.into());
        assert_eq!(params.whoisreply, Some("Some raw WHOIS data".into()));
    }

    // ── WhoisParams serde ────────────────────────────────────────────────

    #[test]
    fn test_whois_params_serialization() {
        let params = WhoisParams {
            domain: Some("test.com".into()),
            status: Some(DomainStatus::Available),
            registrar: None,
            company: None,
            creation_date: Some("2020-01-01".into()),
            update_date: None,
            expiry_date: None,
            whoisreply: None,
        };
        let json = serde_json::to_string(&params).unwrap();
        assert!(json.contains("\"domain\":\"test.com\""));
        assert!(json.contains("\"creationDate\":\"2020-01-01\""));
        assert!(json.contains("\"status\":\"available\""));
    }

    #[test]
    fn test_whois_params_deserialization() {
        let json = r#"{
            "domain": "foo.com",
            "status": "unavailable",
            "registrar": "RegCo",
            "company": null,
            "creationDate": "2020-01-01",
            "updateDate": null,
            "expiryDate": "2030-01-01",
            "whoisreply": null
        }"#;
        let params: WhoisParams = serde_json::from_str(json).unwrap();
        assert_eq!(params.domain, Some("foo.com".into()));
        assert_eq!(params.status, Some(DomainStatus::Unavailable));
        assert_eq!(params.registrar, Some("RegCo".into()));
        assert!(params.company.is_none());
    }
}